olive/model/__init__.py

```diff
@@ -637,24 +636,49 @@ def prepare_session(
     ):
         return self.load_model().eval()
 
+    def get_io_config(self) -> Dict[str, Any]:
+        """Return io config of the model.
+
+        Priority: io_config > hf_config (using onnx_config)
+        """
+        if isinstance(self.io_config, dict):
+            # io_config is provided
+            return self.io_config
+
+        if isinstance(self.io_config, (str, Callable)):
+            # io_config is a string name or a callable
+            logger.debug(f"Calling {self.io_config} to get io_config")
+            user_module_loader = UserModuleLoader(self.model_script, self.script_dir)
+            io_config = user_module_loader.call_object(self.io_config, self)
+            return validate_config(io_config, IOConfig).dict()
+
+        if self.hf_config and self.hf_config.task and not self.hf_config.components:
+            # hf_config is provided
+            logger.debug("Using hf onnx_config to get io_config")
+            return self.hf_config.get_io_config(self.model_path)
+
+        return None
+
     def get_dummy_inputs(self):
         """Return a dummy input for the model."""
         if self.dummy_inputs is not None:
             return self.dummy_inputs
 
         # Priority: dummy_inputs_func > io_config.input_shapes > hf_config.dataset > onnx_config
         dummy_inputs = None
+        io_config = self.get_io_config()
         if self.dummy_inputs_func is not None:
             logger.debug("Using dummy_inputs_func to get dummy inputs")
             user_module_loader = UserModuleLoader(self.model_script, self.script_dir)
             dummy_inputs = user_module_loader.call_object(self.dummy_inputs_func, self)
-        elif self.io_config and self.io_config["input_shapes"]:
+        elif self.io_config and io_config["input_shapes"]:
```

Comment:
do we need this check?

Reply:
Yes, there is a comment below about checking `self.io_config`. We don't want to use io_config that is extracted from hf_config. 

Although, this can also be replaced by `elif io_config.get["input_shapes"]` since the io_config from hf_config never has `input_shapes`. https://github.com/microsoft/Olive/blob/main/olive/model/hf_utils.py#L362 

Depends on whether we want the logic to be clear or simpler.

Reply:
the check seems a little bit hacky since it is not used in the subsequent code.

The following line seems deadcode?

        if self.hf_config and self.hf_config.task and not self.hf_config.components:
            # hf_config is provided
            logger.debug("Using hf onnx_config to get io_config")
            return self.hf_config.get_io_config(self.model_path)

Reply:
It seems we need split the the get_io_config into two different methods. Once is only get from self.io_config, another is only get from hf_config. 

From the time being, the method has two responsibility, which violates the [https://en.wikipedia.org/wiki/Single-responsibility_principle](https://en.wikipedia.org/wiki/Single-responsibility_principle)

Reply:
It is not dead code. This part is used if there is no `io_config` but we have `hf_config`. Just added a test case for this. 

This is also the reason we check for `self.io_config` since `hf_config` is used to generate io_config in a later `elif` clause.

Reply:
I don't think it is doing two responsibilities. It is doing the same thing as `get_dummy_inputs` where the `dummy_inputs` or `io_config` for a model can originate from different sources. The `hf_config` cases are special cases.

Reply:
I mean deadcode in get_dummy_inputs. This kind of behavior seems weird. We'd better refactor code to make sure the logic clear.

Reply:
it's not deadcode because we check io_config from `hf_config` doesn't have shapes.

Reply:
If it has only one responsibility, then the check for self.io_config will be needless. the check seems redundant if we don't look into how get_io_config is implemented.

Reply:
I don't see where anything is weird or unclear. The comments clearly say what the priority of io_config and dummy_inputs are. 
We already handle hf_config as special cases in model loading, etc. It is part of the pytorch model's config can be used for model loading, io_config, dummy_inputs etc so that the user doesn't need to provide them. 

We are just abstracting the different ways of getting the model, io_config, etc under the `load_model`, `get_dummy_inputs`, etc methods.
These logics need to be handled somewhere anyways. Better to do it in the model object itself.

Reply:
Updated as per offline discussion.